internal/queue/pending_txs.go

```diff
@@ -24,30 +23,49 @@ func NewPendingTxs(logger *logger.SugarLogger) *PendingTxs {
 	}
 }
 
-func (p *PendingTxs) Add(txID string, promise *CompletionPromise) {
-	p.Lock()
-	defer p.Unlock()
+// Add returns true if the txId was already taken
+func (p *PendingTxs) Add(txID string, promise *CompletionPromise) bool {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	_, loaded := p.txs[txID]
+	if !loaded {
+		p.txs[txID] = promise
+	}
+	return loaded
+}
+
+func (p *PendingTxs) DeleteWithNoAction(txID string) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	delete(p.txs, txID)
+}
 
-	p.txs[txID] = promise
+func (p *PendingTxs) loadAndDelete(txID string) (*CompletionPromise, bool) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	promise, loaded := p.txs[txID]
+	if loaded {
+		delete(p.txs, txID)
+	}
+	return promise, loaded
 }
 
 // DoneWithReceipt is called after the commit of a block.
 // The `txIDs` slice must be in the same order that transactions appear in the block.
 func (p *PendingTxs) DoneWithReceipt(txIDs []string, blockHeader *types.BlockHeader) {
 	p.logger.Debugf("Done with receipt, block number: %d; txIDs: %v", blockHeader.GetBaseHeader().GetNumber(), txIDs)
 
-	p.Lock()
-	defer p.Unlock()
-
 	for txIndex, txID := range txIDs {
-		p.txs[txID].done(
+		promise, loaded := p.loadAndDelete(txID)
+		if !loaded {
+			continue
+		}
+		promise.done(
```

Comment:
Does it make sense to take the lock len(txIDs) times instead of once? 
We could also `loadAndDelete` all the promises into a slice `[]promise` (known size, can pre-allocate)  under the lock and then call `done` in a loop out side the lock.  
We want to minimize the total time the lock is taken and the number of times it is taken.

Reply:
There is a tradeoff here. Locking once will improve the commit time but will slow down the insert-to-queue time.
Since the number of TXs can be 1e3 up to 1e6, I'm not sure what is the best approach.
A way to avoid the commit time wasted on locking will be to invoke the promises on a non-blocking go routine.
I.e., call `internal/blockprocessor/processor.go:136` inside a go routine without waiting for it.

Reply:
We definitely want to reduce the commit time. 
So I think that is a good idea - delegate the release of the promises to a goroutine. I still think that goroutine should take the lock once for every block, not once for every tx. This will reduce contention for the lock and will reduce the CPU overall.

Reply:
Following is a micro-benchmark of 50:50 insert-if-absent/delete-if-present. The delete-if-present operations are batched, i.e., multiple keys are deleted in the same go routine, while each insert-if-absent is executed in its own go routine.

I compared four implementations: `sync.map`, [`xsync.map`](https://github.com/puzpuzpuz/xsync), regular map with a mutex, and regular map with a mutex that is locked only once per delete batch.

The X axis is the size of the delete-batch, i.e., how many keys are deleted in a single batch.
The Y axis is the operation time (nanoseconds).

This shows that batch locking can reduce the operation's time, but the use of `xsync` can improve the performance significantly.

![image](https://user-images.githubusercontent.com/1833758/218310092-4fd16c64-7f44-411c-bbc6-dad5e00b2ba0.png)